### PR TITLE
Makes crematorium actually burn stuff instead of deleting it

### PR DESCRIPTION
**Pull request by XDTM**
## About The Pull Request

Changes the behaviour of the crematorium.
Now, instead of instantly deleting all its contents, it will process while it remains on.

On every process tick, it will deal 90 burn damage (25 with the RESISTHEAT trait) and 10 firestacks (+ignite) to all mobs, and 50 burn damage(+regular burning) to all other non-fireproof objects.
On top of that mobs will have a greatly accelerated limb cremation process compared to regular burning, although they still need to be on fire for the actual dusting to take place.

**Note: i've not been able to compile or test this yet, putting the PR up early for feedback/opinions/kneejerk reactions. Will test as soon as i can work on a pc with byond on.**

## Why It's Good For The Game

Allows the crematorius to interact more consistently, especially when facing other game mechanics.
You could cremate something for three seconds and have it come out somewhat intact, although still badly burned (and still burning). With enough time everything will burn to ash anyway, of course, but you can now enjoy hearing every bit crumbling into dust inside the crematorium.

Fire immunity now prevents the crematorium from fully dusting a body, although the extreme heat will still deal some burn damage.

As another bonus, after the cremation you can easily collect any fireproof gear left over from the victims!

## Changelog
:cl: XDTM
balance: Crematoriums no longer instantly delete their contents upon activation. Instead, they can be toggled on or off, and will rapidly burn their contents into ash over time while active, unless completely fireproof.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/morgue.dm [hunk 1]
```diff
@@ -242,67 +247,69 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
+	if(!cremating && set_cremate != FALSE)
+		if(user)
+			last_user = user
+		else
+			last_user = null //let's not blame the previous guy
+		cremating = TRUE
+		first_cremation_tick = TRUE
 		locked = TRUE
+		START_PROCESSING(SSobj, src)
 		update_icon()
```
Comment (bobbahbrown) [inline code/game/objects/structures/morgue.dm:278]: are the cases backwards for the audible messages? looks like you're doing the turn off one when it turns on and vice versa

also why do ``set_cremate != FALSE`` and same for true, why not just ``set_cremate`` and ``!set_cremate``?

also also if !user is going to be null, why not just last_user = user

File: code/game/objects/structures/morgue.dm [hunk 2]
```diff
@@ -242,67 +247,69 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
-		IC.set_cone_type("waffle")
-		IC.add_mob_flavor(i_scream)
-		icecreams += IC
-	. = ..()
-	for(var/obj/IC in icecreams)
-		IC.forceMove(src)
+///Ignites and causes heavy burn damage to mobs and objects inside. Non-immune mobs will quickly be incinerated.
+/obj/structure/bodycontainer/crematorium/proc/cremate()
+	if(!cremating)
+		return
```
Comment (bobbahbrown) [inline code/game/objects/structures/morgue.dm:286]: you already checked this in process

File: code/game/objects/structures/morgue.dm [hunk 3]
```diff
@@ -242,67 +247,69 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
+		if(iscarbon(L))
+			var/mob/living/carbon/C = L
+			C.adjust_cremation(25)
+
+	for(var/obj/O in conts) //conts defined above, ignores crematorium and tray
+		if(!(O.resistance_flags & FIRE_PROOF))
+			O.fire_act()
+			O.take_damage(50, BURN, null, FALSE)
+
+	first_cremation_tick = FALSE
```
Comment (bobbahbrown) [inline code/game/objects/structures/morgue.dm:312]: you could probably assign this in the if where first_cremation_tick matters, no need to assign every process call

File: code/modules/assembly/doorcontrol.dm [hunk 4]
```diff
@@ -143,6 +143,6 @@
 	cooldown = TRUE
 	for (var/obj/structure/bodycontainer/crematorium/C in GLOB.crematoriums)
 		if (C.id == id)
-			C.cremate(usr)
+			C.toggle_cremation(usr)
```
Comment (bobbahbrown) [inline code/modules/assembly/doorcontrol.dm:146]: this doesn't work with the signature you made for toggle_cremation

**Comment by wesoda25** [comment]
Seems really cool

File: code/modules/mob/living/silicon/damage_procs.dm [hunk 5]
```diff
@@ -1,5 +1,5 @@
 
-/mob/living/silicon/apply_damage(damage = 0,damagetype = BRUTE, def_zone = null, blocked = FALSE, forced = FALSE)
+/mob/living/silicon/apply_damage(damage = 0,damagetype = BRUTE, def_zone = null, blocked = FALSE, forced = FALSE, spread_damage = FALSE)
```
Comment (bobbahbrown) [inline code/modules/mob/living/silicon/damage_procs.dm:2]: why did you add this

**Comment by Time-Green** [comment]
Since slimes heal from burn damage, you can turn a crematorium into a pet slime medbay

👍

**Comment by Pugie** [comment]
But why do this when it stops cremator from cremating? Cremating bodies with fire hardly ever works... For whatever reason limbs can be fully burnt/mangled and not register as mangled/peeling away which stops them from crumbling to ash & halting the entire cremation process. Fire cremation has been buggy and seldom worked ever since it was introduced. That bug should be fixed first.

**Comment by SkeletalElite** [comment]
How long will this take to cremate a body? Depending on how long this is this could be a little buff to lings, although you can still shuttlegib them, but thats still one less way to kill them.

File: code/game/objects/structures/morgue.dm (hunk 1 above)
Comment (XDTM) [reply code/game/objects/structures/morgue.dm:278]: 1) they're backwards, will fix
2) my intention was to essentially make it skip the activation/deactivation is set_cremate would prevent it, while also ignoring the value if it's null. I'm not sure right now if `!= FALSE` would also catch null though, that would break the check.
3) As the comment implies, it makes sure that if user A cremates someone and then the crematorium is activated again later without an user it doesn't get logged as A doing it

File: code/game/objects/structures/morgue.dm (hunk 2 above)
Comment (XDTM) [reply code/game/objects/structures/morgue.dm:286]: i opted to have more checks to be safe, but they can be removed if redundancy is a problem

File: code/game/objects/structures/morgue.dm (hunk 3 above)
Comment (XDTM) [reply code/game/objects/structures/morgue.dm:312]: right, will do

File: code/modules/assembly/doorcontrol.dm (hunk 4 above)
Comment (XDTM) [reply code/modules/assembly/doorcontrol.dm:146]: will fix

File: code/modules/mob/living/silicon/damage_procs.dm (hunk 5 above)
Comment (XDTM) [reply code/modules/mob/living/silicon/damage_procs.dm:2]: linter complained about the spread_damage argument not being there, so i added it. Might also prevent issues down the line, anyway.

**Comment by XDTM** [comment]
Reporting those bugs would help in getting them fixed. 

Though odds are that the reason that limbs aren't burning off in those cases is because either they aren't on fire, they don't have enough cremation_counter yet (it takes a while), or the chest isn't at max damage.

I initially made cremation very forgiving since it was effectively body disposal without any external requirements, but it _could_ be made more aggressive to make it more reliable to use.

File: code/game/objects/structures/morgue.dm (hunk 1 above)
Comment (bobbahbrown) [reply code/game/objects/structures/morgue.dm:278]: on 2 I can see where you're coming from I suppose

on 3 if its activated without a user won't user be null and hence you can just set last_user to user (which should be null in that case?)

Comment (XDTM) [reply code/game/objects/structures/morgue.dm:278]: oh, true, didn't consider that

**Comment by Pugie** [comment]
> Reporting those bugs would help in getting them fixed.
> 
> Though odds are that the reason that limbs aren't burning off in those cases is because either they aren't on fire, they don't have enough cremation_counter yet (it takes a while), or the chest isn't at max damage.
> 
> I initially made cremation very forgiving since it was effectively body disposal without any external requirements, but it _could_ be made more aggressive to make it more reliable to use.

Chest can be at max burn and display the peeling away message but the body will never fully cremate because even one limb will have full damage yet not display the peeling away text, & the only way to fix this is to dismember the limb or heal it and hope having max damage actually registers properly next time you set the limb on fire.

**Comment by XDTM** [comment]
I'll look into that as well, then

**Comment by MrDoomBringer** [comment]
if i cremate a body with a bunch of grenades in their bag, would they detonate? Still, cool PR.

**Comment by XDTM** [comment]
grenades do indeed blow up if they're cremated

**Comment by ShibaInuLord** [comment]
will this make it so high-risk items can't be cremated anymore?

**Comment by XDTM** [comment]
only if they're otherwise fireproof, which is the case for the nuke disk for example

**Comment by AdamElTablawy** [comment]
guess lings are getting put in the gibber and not the crematorium

**Comment by wesoda25** [comment]
> grenades do indeed blow up if they're cremated

Epic ling strat identified

**Comment by Shadowflame909** [comment]
> > grenades do indeed blow up if they're cremated
> 
> Epic ling strat identified

but then you lose the cremator bro...

Unless fire grenades trigger in the crematorium?

**Comment by MrDoomBringer** [comment]
@Shadowflame909 I think he meant that as a ling you would hide grenades in a box or something such that when you got cremated, they would detonate and kill the cremator and those around you as a final dunk

**Comment by ATH1909** [comment]
Tfw the deep fryer is a more reliable way to dispose of items than the crematorium.

**Comment by Shadowflame909** [comment]
Yeah also are handcuffs fireproof

Imagine cremating a ling, and his handcuffs burn off before he revival stasis's Allowing him to self open and start attacking.

Lings just need a rework in general tbh. Still waiting on that @Armhulen

**Comment by Time-Green** [comment]
> Yeah also are handcuffs fireproof
> 
> Imagine cremating a ling, and his handcuffs burn off before he uses revival stasis. Allowing him to self-open and start attacking.
> 
> Lings just need a rework in general tbh. Still waiting on that @Armhulen
> 
> Edit: Or he waits till all his limbs burn off, and like he's about to get perma-ashed, Then he revival stasis's to get back to 100, instant regenerates and breaks out like that. Effectively nullifying the cremation.

Changelings can’t fleshmend or revive when on fire

**Comment by Shadowflame909** [comment]
didn't that nerf get denied/closed because it wasn't finished

**Comment by XDTM** [comment]
still takes time to resist out, ling should end up in crit far before being able to complete the resist

**Comment by IndieanaJones** [comment]
> 
> 
> > Yeah also are handcuffs fireproof
> > Imagine cremating a ling, and his handcuffs burn off before he uses revival stasis. Allowing him to self-open and start attacking.
> > Lings just need a rework in general tbh. Still waiting on that @Armhulen
> > Edit: Or he waits till all his limbs burn off, and like he's about to get perma-ashed, Then he revival stasis's to get back to 100, instant regenerates and breaks out like that. Effectively nullifying the cremation.
> 
> Changelings can’t fleshmend or revive when on fire

This isn't true as of current.  Whether or not it's intentional, I don't know.

**Comment by Shadowflame909** [comment]
Even worse is that Lings have counters for LITERALLY ANYTHING. (Who gave them all these counters)

They have that "acid cuffs" technique that'll dissolve cuffs in a second, so you might be cucked literally as soon as you put them in the crematorium.

**Comment by XDTM** [comment]
the crematorius has its own resist timer when shut, cuffs or not

File: code/game/objects/structures/morgue.dm [hunk 6]
```diff
@@ -242,67 +247,65 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
 	return
 
-/obj/structure/bodycontainer/crematorium/proc/cremate(mob/user)
-	if(locked)
-		return //don't let you cremate something twice or w/e
-	// Make sure we don't delete the actual morgue and its tray
-	var/list/conts = GetAllContents() - src - connected
+/obj/structure/bodycontainer/crematorium/process()
+	if(cremating)
+		cremate()
```
Comment (Qustinnus) [inline code/game/objects/structures/morgue.dm:257]: you might as well handle cremation in this proc as well, it's a bit of a buzzword but proc overhead is still an issue.

File: code/game/objects/structures/morgue.dm [hunk 7]
```diff
@@ -242,67 +247,65 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
+		L.IgniteMob()
+
+		//Artificially speed up the cremation process
+		if(iscarbon(L))
+			var/mob/living/carbon/C = L
+			C.adjust_cremation(25)
+
+	for(var/obj/O in conts) //conts defined above, ignores crematorium and tray
+		if(!(O.resistance_flags & FIRE_PROOF))
+			O.fire_act()
```
Comment (Qustinnus) [inline code/game/objects/structures/morgue.dm:307]: This proc actually takes a temperature var which is very rarely used, but should still be filled out.

File: code/game/objects/structures/morgue.dm [hunk 8]
```diff
@@ -242,67 +247,65 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
-		audible_message("<span class='hear'>You hear a roar as the crematorium activates.</span>")
+///Toggles cremation. The cremate parameter can be used to set a specific value instead of toggling.
+/obj/structure/bodycontainer/crematorium/proc/toggle_cremation(set_cremate, mob/user)
+	if(cremating && set_cremate != TRUE)
+		cremating = FALSE
+		locked = FALSE
+		STOP_PROCESSING(SSobj, src)
+		update_icon()
 
+		audible_message("<span class='hear'>You hear a roar as [src] activates.</span>")
```
Comment (Qustinnus) [inline code/game/objects/structures/morgue.dm:267]: this should be in the block below

File: code/game/objects/structures/morgue.dm [hunk 9]
```diff
@@ -242,67 +247,65 @@ GLOBAL_LIST_EMPTY(crematoriums)
...
-
-		if(!QDELETED(src))
-			locked = FALSE
-			update_icon()
-			playsound(src.loc, 'sound/machines/ding.ogg', 50, TRUE) //you horrible people
-
-/obj/structure/bodycontainer/crematorium/creamatorium
-	name = "creamatorium"
-	desc = "A human incinerator. Works well during ice cream socials."
+		audible_message("<span class='hear'>The roar of [src]'s flames gradually dies down.</span>")
```
Comment (Qustinnus) [inline code/game/objects/structures/morgue.dm:277]: and vice versa

**Comment by Qustinnus** [comment]
why tho

**Comment by XDTM** [comment]
haven't been able to spend time on coding lately, and it was about to go stale

 i'll reopen it when i can actually test it
